source/v1/I2C.cpp

```diff
@@ -0,0 +1,229 @@
+ * limitations under the License.
+ */
+#include "mbed-drivers/v1/I2C.hpp"
+#include "minar/minar.h"
+#include "ualloc/ualloc.h"
+#include "core-util/CriticalSectionLock.h"
+#include "PeripheralPins.h"
+
+#if DEVICE_I2C && DEVICE_I2C_ASYNCH
+
+namespace mbed_drivers {
+using namespace mbed;
+namespace v1 {
+I2CTransaction::I2CTransaction(uint16_t address, uint32_t hz, bool irqsafe, I2C * issuer):
+    next(nullptr),
+    address(address),
+    root(nullptr),
+    current(nullptr),
+    hz(hz),
+    repeated(false),
+    irqsafe(irqsafe),
+    issuer(issuer)
+{}
+I2CTransaction::~I2CTransaction() {
+    current = root;
+    while (current) {
+        detail::I2CSegment * next = current->get_next();
+        issuer->free(current, irqsafe);
+        current = next;
+    }
+}
+
+detail::I2CSegment * I2CTransaction::new_segment()
+{
+    detail::I2CSegment * s = issuer->new_segment(irqsafe);
+    CORE_UTIL_ASSERT(s != nullptr);
+    if (!s) {
+        return nullptr;
+    }
+    s->set_next(nullptr);
+    mbed::util::CriticalSectionLock lock;
+    if (root == NULL) {
+        root = s;
+        current = s;
+    } else {
+        current->set_next(s);
+        current = s;
+    }
+    return s;
+}
+
+I2C::I2C(PinName sda, PinName scl) :
+    _hz(100000)
+{
+    // Select the appropriate I2C Resource Manager
+    uint32_t i2c_sda = pinmap_peripheral(sda, PinMap_I2C_SDA);
+    uint32_t i2c_scl = pinmap_peripheral(scl, PinMap_I2C_SCL);
+    ownerID = pinmap_merge(i2c_sda, i2c_scl);
+    int rc = detail::get_I2C_owner(ownerID)->init(sda, scl);
+    if (rc) {
+        ownerID = (size_t) -1;
+    }
+}
+
+void I2C::frequency(int hz)
+{
+    _hz = hz;
+}
+I2C::TransferAdder I2C::transfer_to(int address) {
+    TransferAdder t(this, address, _hz, false);
+    return t;
+}
+
+int I2C::post_transaction(I2CTransaction * t) {
+    if (ownerID == (size_t) -1) {
+        return -1;
+    }
+    return detail::get_I2C_owner(ownerID)->post_transaction(t);
+}
+
+detail::I2CSegment * I2C::new_segment(bool irqsafe) {
+    detail::I2CSegment * newseg = nullptr;
+    if (irqsafe) {
+        if (!SegmentPool) {
+            return nullptr;
+        }
+        void * space = SegmentPool->alloc();
+        if (!space) {
+            return nullptr;
+        }
+        newseg = new(space) detail::I2CSegment();
+    } else {
+        newseg = new detail::I2CSegment();
+    }
+    return newseg;
+}
+
+I2CTransaction * I2C::new_transaction(uint16_t address, uint32_t hz, bool irqsafe, I2C * issuer)
+{
+    I2CTransaction * t;
+    if (irqsafe) {
+        if (!TransactionPool) {
+            return nullptr;
+        }
+        void *space = TransactionPool->alloc();
+        if (!space) {
+            return nullptr;
+        }
+        t = new(space) I2CTransaction(address, hz, irqsafe, issuer);
+    } else {
+        t = new I2CTransaction(address, hz, irqsafe, issuer);
+    }
+    return t;
+}
+void I2C::free(detail::I2CSegment * s, bool irqsafe)
```

Comment:
Looking through this pull request there are sequences of function (and member) definitions (and declarations) where there are no intervening blank lines, some instances with one intervening blank line and others with two intervening blank lines.  Can we adopt a level of conformity with the rest of the code base and aim for 1 blank line in each instance?